drivers/pinctrl/pinctrl_nrf.c

```diff
@@ -73,415 +35,657 @@ static const nrf_gpio_pin_drive_t drive_modes[NRF_DRIVE_COUNT] = {
-			dir = NRF_GPIO_PIN_DIR_INPUT;
-			input = NRF_GPIO_PIN_INPUT_DISCONNECT;
-			break;
-#endif /* DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_exmif) */
 #if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_twis) || defined(CONFIG_NRFX_TWIS)
+	nrf_pinctrl_twis,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_i2s) || defined(CONFIG_NRFX_I2S)
+	nrf_pinctrl_i2s,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_pdm) || defined(CONFIG_NRFX_PDM)
+	nrf_pinctrl_pdm,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_pwm) || defined(CONFIG_NRFX_PWM)
+	nrf_pinctrl_pwm,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_qdec) || defined(CONFIG_NRFX_QDEC)
+	nrf_pinctrl_qdec,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_qspi) || defined(CONFIG_NRFX_QSPI)
+	nrf_pinctrl_qspi,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_exmif)
+	nrf_pinctrl_exmif,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_can)
+	nrf_pinctrl_can,
+#endif
+#if DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_grtc) || defined(CONFIG_NRFX_GRTC)
+	nrf_pinctrl_grtc,
+#endif
+};
 
-		case NRF_FUN_TWIS_SCL:
-			nrf_twis_scl_pin_set((NRF_TWIS_Type *)reg, psel);
-			if (drive == NRF_GPIO_PIN_S0S1) {
-				drive = NRF_GPIO_PIN_S0D1;
-			}
-			dir = NRF_GPIO_PIN_DIR_INPUT;
-			input = NRF_GPIO_PIN_INPUT_CONNECT;
-			break;
-		case NRF_FUN_TWIS_SDA:
-			nrf_twis_sda_pin_set((NRF_TWIS_Type *)reg, psel);
-			if (drive == NRF_GPIO_PIN_S0S1) {
-				drive = NRF_GPIO_PIN_S0D1;
-			}
-			dir = NRF_GPIO_PIN_DIR_INPUT;
-			input = NRF_GPIO_PIN_INPUT_CONNECT;
-			break;
-#endif /* DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_twis) || defined(CONFIG_NRFX_TWIS) */
-		default:
+int pinctrl_configure_pins(const pinctrl_soc_pin_t *pins, uint8_t pin_cnt,
+			   uintptr_t reg)
+{
+#ifdef CONFIG_SOC_NRF54H20_GPD
+	bool gpd_requested = false;
+#endif
+
+	for (uint8_t i = 0U; i < pin_cnt; i++) {
+		struct nrf_pinctrl_pin pin = {
+			.reg              = reg,
+			.pin_num          = NRF_GET_PIN(pins[i]),
+			.pull             = NRF_GET_PULL(pins[i]),
+			.drive            = NRF_GET_DRIVE(pins[i]),
+			.low_power        = NRF_GET_LP(pins[i]),
+			.invert           = NRF_GET_INVERT(pins[i]),
+			.clock_pin_en     = NRF_GET_CLOCKPIN_ENABLE(pins[i]),
+			.gpd_fast_active1 = NRF_GET_GPD_FAST_ACTIVE1(pins[i]),
+			.fun              = NRF_GET_FUN(pins[i]),
+		};
+
+		if (pin.drive >= ARRAY_SIZE(drive_modes)) {
+			return -EINVAL;
+		}
+
+		struct nrf_pinctrl_config cfg = {
+			.write        = NRF_PINCTRL_WRITE_NONE,
+			.drive        = drive_modes[pin.drive],
+			.sense        = NRF_GPIO_PIN_NOSENSE,
+			.clock_pin_en = pin.clock_pin_en,
+			.pull         = pin.pull,
+		};
+
+		if (pin.pin_num == NRF_PIN_DISCONNECTED) {
+			pin.pin_num = PSEL_DISCONNECTED;
+		}
+
+		/* Iterate over pinctrl functions array.
+		 * If pin is handled by a given function, it returns true.
+		 */
+		bool handled = false;
+		uint8_t fun_idx = 0;
+
+		while (!handled && fun_idx < ARRAY_SIZE(funs_table)) {
+			handled = funs_table[fun_idx](&pin, &cfg);
+			fun_idx++;
+		}
```

Comment:
I don't really understand the reason for this change. It is not required for introduction of the SoC-specific hooks and iterating over this array of functions is much less efficient than that "huge switch-case" that was here before. In the worst case, all these functions must be called before a pin is configured for the least lucky peripheral (and this is repeated for all the pins configured for a given peripheral). What's the benefit?

Reply:
This will allow using pinctrl functions provided by nrfx drivers.

Reply:
For peripherals supported on driver level, nrfx would expose API compatible with `nrf_pinctrl_fun` that replaces custom implementation in pincontrol SHIM.

Reply:
Such refactoring can be done then, if it indeed turns out to be beneficial. I don't see the point in doing this now.